ciao/arch/architecture.md

```diff
@@ -0,0 +1,423 @@
+later date.
+
+Instances, workloads and images are created inside constructs called tenants.
+While the instance is the unit of computation, the tenant is the unit of isolation.
+Instances running in the same tenant are connected together on a tenant private
+network.  Instances that are part of the same tenant can talk to each other but
+those that are part of different tenants, by default, cannot.  This holds true
+even if two instances from different tenants happen to be running on the same
+compute node.  Workloads and images are also, by default, private to the tenants
+in which they were created, although it is possible to create public workloads
+and images that can be used and shared across all tenants.
+
+If the cluster is connected to an external network, instances will be able to
+access services running in that external network.  Although, external client
+applications will not be able to see or talk to instances by default, instances
+can be exposed to the outside world by assigning external IP addresses to them,
+as we shall see below.
+
+Once launched, an instance can be stopped and restarted.  Stopped instances
+typically lose the association with the node on which they previously ran.  By
+stopping and restarting an instance you migrate it from
+one node to another.  Any data stored in data volumes, and for VMs any data
+stored in rootfs volumes, will be migrated intact with the instance.  Live
+migration is not supported.  Entire nodes can be evacuated of all of their
+instances in a single command.  This is useful if a node needs to be temporarily
+removed from the cluster for maintenance.
+
+Two levels of privilege are defined; user level and administrator level.  Users can
+perform many actions including creating and destroying instances, images and volumes.
+However, certain actions are restricted to administrators.  Examples of operations
+that require administrator level access include creating public workloads and images
+and evacuating nodes.
+
+Finally, quotas can be applied to a tenant by cluster administrators limiting
+the number of instances, volumes, images and resources that can be consumed by
+that tenant.
+
+## Nodes and services
+
+A Cloud Integrated Advanced Orchestrator cluster is composed of one or more
+nodes.  These nodes host the cluster's control plane and its instances.  Three
+different types of nodes are defined; master, compute and network.  The control
+plane consists of four different types of services; controller, scheduler, launcher
+and cnci-agent.
+
+Controller and scheduler run on the master node.  Controller maintains
+the cluster state and provides a public REST API that can be used to construct
+and maintain the cluster.  Scheduler acts as a central hub for the cluster,
+routing control plane traffic to and from its various services.  At the time of
+writing, only one instance of both controller and scheduler run at any one time.
```

Comment:
It may be useful to highlight that the scheduler can crash/restart and rebuild state without using a persistent store.

Reply:
Another good point.  I'll mention this as well.